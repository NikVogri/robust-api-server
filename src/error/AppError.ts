interface IAppError {
    message: string;
    statusCode: number;
    thrownError?: Error | unknown;
    payload?: Record<string, unknown>;
}

export class AppError extends Error {
    thrownError?: Error | unknown;
    statusCode: number;
    payload?: Record<string, unknown>;

    constructor(public data: IAppError) {
        super(data.message);

        this.thrownError = data.thrownError;
        this.statusCode = data.statusCode;
        this.payload = data.payload;
    }

    static toStringifiableObject(error: Error | unknown) {
        if (!(error instanceof Error)) return {};

        const err: Record<string, unknown> = {};

        Object.getOwnPropertyNames(error).forEach((prop) => {
            if ((error[prop as keyof Error] as unknown) instanceof Error) {
                err[prop] = AppError.toStringifiableObject(error);
            } else err[prop] = error[prop as keyof Error];
        });

        return err;
    }
}
