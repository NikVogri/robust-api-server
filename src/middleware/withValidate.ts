import { NextFunction, Request, Response } from 'express';
import { ZodError, ZodSchema } from 'zod';
import { AppError } from '../error/AppError';

interface RequestSchema {
    body?: ZodSchema;
    params?: ZodSchema;
    query?: ZodSchema;
}

interface ParseErrors {
    schema: keyof RequestSchema;
    error: ZodError;
}

export const withValidate = (schema: RequestSchema) => async (req: Request, _: Response, next: NextFunction) => {
    const parseErrors: ParseErrors[] = [];

    if (schema.params) {
        const parsed = schema.params.safeParse(req.params);
        if (!parsed.success) parseErrors.push({ schema: 'params', error: parsed.error });
    }

    if (schema.query) {
        const parsed = schema.query.safeParse(req.query);
        if (!parsed.success) parseErrors.push({ schema: 'query', error: parsed.error });
    }

    if (schema.body) {
        const parsed = schema.body.safeParse(req.body);
        if (!parsed.success) parseErrors.push({ schema: 'body', error: parsed.error });
    }

    if (parseErrors.length > 0) {
        next(
            new AppError({
                message: 'Invalid payload',
                statusCode: 400,
                payload: {
                    errors: parseErrors.reduce((curr, acc) => {
                        return {
                            ...curr,
                            [acc.schema]: acc.error.issues,
                        };
                    }, {}),
                },
            }),
        );
    } else {
        next();
    }
};
