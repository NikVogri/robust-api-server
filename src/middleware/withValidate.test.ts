import { describe, expect, it, vi } from 'vitest';
import { withValidate } from './withValidate';
import z from 'zod';
import { getRequest, getResponse } from '../../vitest.mock';
import { AppError } from '../error/AppError';

describe('Test withValidate', () => {
    it('should should go to the next middleware without forwarding an error when all validation', async () => {
        expect.assertions(1);

        const request = getRequest();
        const response = getResponse();
        const nextFn = vi.fn();

        nextFn.mockImplementationOnce((...args) => {
            expect(args[0]).toBeUndefined();
            return undefined;
        });

        const schemaValidator = withValidate({
            params: z.object({ id: z.string() }),
            body: z.object({ username: z.string() }),
            query: z.object({ page: z.string() }),
        });

        request.query = { page: '1' };
        request.body = { username: 'username' };
        request.params = { id: 'identifier' };

        await schemaValidator(request, response, nextFn);
    });

    it('should forward error when parameter validation fails', async () => {
        expect.assertions(2);

        const request = getRequest();
        const response = getResponse();
        const nextFn = vi.fn();

        nextFn.mockImplementationOnce((...args) => {
            expect(args[0]).toBeInstanceOf(AppError);
            expect(JSON.stringify(args[0])).toMatchSnapshot();
            return undefined;
        });

        const schemaValidator = withValidate({
            params: z.object({
                id: z.string(),
            }),
        });

        await schemaValidator(request, response, nextFn);
    });

    it('should forward error when query validation fails', async () => {
        expect.assertions(2);

        const request = getRequest();
        const response = getResponse();
        const nextFn = vi.fn();

        nextFn.mockImplementationOnce((...args) => {
            expect(args[0]).toBeInstanceOf(AppError);
            expect(JSON.stringify(args[0])).toMatchSnapshot();
            return undefined;
        });

        const schemaValidator = withValidate({
            query: z.object({
                page: z.number(),
            }),
        });

        await schemaValidator(request, response, nextFn);
    });

    it('should forward error when body validation fails', async () => {
        expect.assertions(2);

        const request = getRequest();
        const response = getResponse();
        const nextFn = vi.fn();

        nextFn.mockImplementationOnce((...args) => {
            expect(args[0]).toBeInstanceOf(AppError);
            expect(JSON.stringify(args[0])).toMatchSnapshot();
            return undefined;
        });

        const schemaValidator = withValidate({
            body: z.object({
                username: z.string(),
            }),
        });

        await schemaValidator(request, response, nextFn);
    });
});
