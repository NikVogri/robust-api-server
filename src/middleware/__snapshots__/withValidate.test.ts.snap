// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test withValidate > should forward error when body validation fails 1`] = `"{"data":{"message":"Invalid payload","statusCode":400,"payload":{"errors":{"body":[{"code":"invalid_type","expected":"object","received":"undefined","path":[],"message":"Required"}]}}},"statusCode":400,"payload":{"errors":{"body":[{"code":"invalid_type","expected":"object","received":"undefined","path":[],"message":"Required"}]}}}"`;

exports[`Test withValidate > should forward error when parameter validation fails 1`] = `"{"data":{"message":"Invalid payload","statusCode":400,"payload":{"errors":{"params":[{"code":"invalid_type","expected":"object","received":"undefined","path":[],"message":"Required"}]}}},"statusCode":400,"payload":{"errors":{"params":[{"code":"invalid_type","expected":"object","received":"undefined","path":[],"message":"Required"}]}}}"`;

exports[`Test withValidate > should forward error when query validation fails 1`] = `"{"data":{"message":"Invalid payload","statusCode":400,"payload":{"errors":{"query":[{"code":"invalid_type","expected":"object","received":"undefined","path":[],"message":"Required"}]}}},"statusCode":400,"payload":{"errors":{"query":[{"code":"invalid_type","expected":"object","received":"undefined","path":[],"message":"Required"}]}}}"`;
